Migrate HubSpot Marketing Emails stream from v1 to v3 API

katmarkham opened the pull request:
## What
- Replace deprecated /marketing-emails/v1/emails/with-statistics endpoint
- Significant schema change for `marketing_emails` stream necessitates a breaking change
- Update response field path from 'objects' to 'results' for v3 API compatibility
- Update unit tests to mock v3 API endpoints
- Bump version to 6.0.0 with breaking change documentation
- Add upgrade deadline of October 1, 2025 per HubSpot's sunset timeline

Fixes upcoming API deprecation: https://developers.hubspot.com/changelog/marketing-email-api-v3-released-to-general-availability-and-upcoming-sunset-for-v1

Closes: https://github.com/airbytehq/airbyte-internal-issues/issues/12895

## Review guide
1. manifest.yaml
2. tests
3. `metadata.yaml` & `hubspot-migrations.md`

## User Impact
- Breaking change!

## Can this PR be safely reverted and rolled back?
<!--
* If unsure, leave it blank.
-->
- [ ] YES 💚
- [x] NO ❌

katmarkham committed:
Migrate HubSpot Marketing Emails stream from v1 to v3 API

- Replace deprecated /marketing-emails/v1/emails/with-statistics endpoint
- Use /marketing/v3/emails for email data and /marketing/v3/emails/{id}/statistics for statistics
- Add MarketingEmailStatisticsTransformation to merge email and statistics data
- Update response field path from 'objects' to 'results' for v3 API compatibility
- Update unit tests to mock v3 API endpoints
- Bump version to 6.0.0 with breaking change documentation
- Add upgrade deadline of October 1, 2025 per HubSpot's sunset timeline

Fixes upcoming API deprecation: https://developers.hubspot.com/changelog/marketing-email-api-v3-released-to-general-availability-and-upcoming-sunset-for-v1

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=65100) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=65100) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=65100) it.</sub>

katmarkham committed:
Add documentation for v6.0.0 Marketing Emails API migration

- Add migration guide for upgrading to 6.0.0 with reset instructions
- Add changelog entry for v6.0.0 explaining API migration from v1 to v3
- Include references to October 1, 2025 HubSpot API sunset deadline

Fixes connector validation errors requiring migration guide and changelog entries

katmarkham committed:
Fix ruff formatting issues

- Clean up whitespace and line formatting in components.py
- Format test_source.py according to project style guidelines
- Ensure code passes ruff formatting checks

katmarkham committed:
Enhance test to verify statistics data is merged into records

- Add assertions to check that statistics fields are present in email records
- Verify specific mocked values (delivered, opens, clicks, bounces, optouts)
- Ensure both email data and statistics data are properly merged
- Add descriptive error messages for each assertion

Addresses code review feedback to validate transformation functionality

katmarkham committed:
Simplify v3 API migration to use includeStats parameter

- Use single /marketing/v3/emails?includeStats=true endpoint instead of complex transformation
- Remove MarketingEmailStatisticsTransformation since stats included in response
- Update all tests to use v3 API endpoints with includeStats parameter
- Improve breaking changes documentation with technical details
- Simplify architecture while maintaining same functionality

katmarkham committed:
Complete merge with master: resolve conflicts and keep HubSpot v3 API improvements

- Resolved Docker tag conflict: keep 6.0.0 (our breaking change version)
- Merged changelog properly: 6.0.0 at top, followed by 5.8.20 and 5.8.19 from master
- Fixed duplicate content in docs changelog
- Maintained our simplified v3 API approach with includeStats=true parameter

The merge brings in all latest master changes while preserving our HubSpot Marketing Emails v3 API migration improvements.

katmarkham committed:
Fix marketing emails test to match simplified v3 API implementation

- Update mock responses to include statistics directly with includeStats=true
- Remove separate statistics endpoint mocks (no longer needed)
- Update test assertions to expect statistics in main response
- Fix test to match our simplified single-call approach vs complex transformation

This fixes the KeyError: 'delivered' test failure by aligning the test with our
actual implementation that uses /marketing/v3/emails?includeStats=true.

katmarkham committed:
Fix ruff formatting issues

- Break up long lines in test functions and comments
- Split long URL constructions into multiple lines
- Fix function signature line length
- Ensure all lines comply with ruff formatting standards

katmarkham committed:
Fix comprehensive ruff formatting violations

- Remove trailing whitespace from marketing emails test data
- Break up long function signatures into multiple lines
- Split long URL constructions and API calls
- Break up long list comprehensions and dictionary structures
- Fix long comments and docstrings
- Break up long assertion lines for readability

All lines now comply with ruff line length requirements (120 chars).

katmarkham commented:
/format-fix

octavia-squidington-iii committed:
chore: auto-fix lint and format issues

pnilan committed:
remove unused definitions

pnilan committed:
update schema, add incremental sync

pnilan committed:
update tests for marketing_streams

pnilan committed:
update migration messaging

pnilan committed:
Merge branch 'master' into hubspot-marketing-emails-v3-migration

pnilan committed:
fix(source-hubspot): updates references to stream schemas (#65601)

tybernstein reviewed (approved):
Kat Approved this

pnilan committed:
update documentation

pnilan committed:
Merge branch 'hubspot-marketing-emails-v3-migration' of github.com:airbytehq/airbyte into hubspot-marketing-emails-v3-migration

brianjlai reviewed (commented):
a few small things and points of discussion, but nothing major to change

i did a quick scan of the schema just to see if anything looked out of the ordinary, but this was presumably auto-generated from an example response?

pnilan commented on airbyte-integrations/connectors/source-hubspot/manifest.yaml:
```diff
@@ -574,21 +577,14 @@ definitions:
         fields:
           - path: ["updated"]
-            value: "{{ record.get('updated') or record['created'] }}"
+            value: "{{ format_datetime(record.get('updatedAt'), '%ms') if record.get('updatedAt') is not none else none }}"
```
And correct, don't need to transform updatedAt.

pnilan committed:
resolve comments

pnilan committed:
remove unneeded test

brianjlai reviewed (approved):
lgtm! looking at the code again, its interesting why we didn't just make this a semi-incremental before because we were sort of filtering on start_date so we couldn't added semi-incremental filtering.

That being said, it actually makes the breaking change easier to manage since users don't have any state that would be lost in the transition
